File: sdk/es/src/stepauth/callback.ts
import { decodeEnvelopePayload } from "./envelope.js";
import { EnvelopeFormatError } from "./crypto.js";
import type { ExecutionResult, Decision } from "./wire.js";
import { verifyDecision, type Config, type PendingState } from "./client.js";

export type { PendingState } from "./client.js";

/** Resolves a requestId to its stored PendingState, or a nullish value if
 * unknown. Return the stored state for any requestId this SP issued, with
 * `result` set once the decision has been executed — never return null for an
 * already-decided id: a null reads as a rejection, and the hub retries a
 * rejection for 24h. The SDK persists nothing; this is the caller's store
 * lookup. */
export type PendingLookup = (
  requestId: string,
) => PendingState | null | undefined | Promise<PendingState | null | undefined>;

/** Invoked with a fully verified decision; returns the ack body to send back
 * to the hub. */
export type DecisionHandler = (decision: Decision) => ExecutionResult | Promise<ExecutionResult>;

/** What createCallbackHandler's handler returns: an HTTP status plus the body
 * to write, so it drops into any server's response object. */
export interface CallbackResponse {
  status: number;
  body: ExecutionResult;
}

// One status for every rejection reason: malformed body, unknown requestId,
// each of verifyDecision's four checks, and a handler throw all come back
// identical. Varying the status by cause would itself be information a
// forged caller could use to find which check to attack next.
const REJECTED_STATUS = 401;

const MAX_CALLBACK_BODY_BYTES = 1 << 20;

function reject(): CallbackResponse {
  return { status: REJECTED_STATUS, body: { requestId: "", status: "error" } };
}

/**
 * Create a framework-agnostic handler for the hub's decision-delivery
 * webhook. The hub POSTs a signed decision to the SP's callbackUrl — there is
 * no polling, and THIS ENDPOINT CARRIES NO TRANSPORT AUTH of its own: the
 * envelope signature verified below is the entire security boundary, so
 * nothing here should special-case an unauthenticated-looking request as
 * lower-risk.
 *
 * Takes the raw body as received (a string, or the bytes before decoding) and
 * returns a status plus the ExecutionResult body; the caller writes both to
 * its actual response.
 */
export function createCallbackHandler(
  cfg: Config,
  lookup: PendingLookup,
  handler: DecisionHandler,
): (body: string | Uint8Array) => Promise<CallbackResponse> {
  return async (body) => {
    const byteLength = typeof body === "string" ? Buffer.byteLength(body) : body.length;
    if (byteLength > MAX_CALLBACK_BODY_BYTES) return reject();

    const raw = typeof body === "string" ? body : Buffer.from(body).toString();

    let requestId: string;
    try {
      const probe = JSON.parse(Buffer.from(decodeEnvelopePayload(raw)).toString()) as {
        requestId?: unknown;
      };
      if (typeof probe.requestId !== "string" || probe.requestId === "") {
        throw new EnvelopeFormatError();
      }
      requestId = probe.requestId;
    } catch {
      return reject();
    }

    // An unknown requestId is rejected BEFORE any verification is attempted:
    // there is nothing to verify a decision against, and verifying first would
    // do cryptographic work on behalf of a caller this SP never solicited.
    let pending: PendingState | null | undefined;
    try {
      pending = await lookup(requestId);
    } catch {
      return reject();
    }
    if (!pending) return reject();

    let decision: Decision;
    try {
      decision = verifyDecision(cfg, raw, pending);
    } catch {
      return reject();
    }

    // Delivery is at-least-once and the hub retries every non-2xx for 24h, so
    // a redelivery of an already-executed decision must replay the recorded
    // result rather than run the handler again.
    if (pending.result) return { status: 200, body: pending.result };

    let result: ExecutionResult;
    try {
      result = await handler(decision);
    } catch (err) {
      // Reaching here means all four verification checks passed, so this is
      // not a forged caller to stonewall: it's the protocol's own "verified,
      // but execution failed" case, and a 401 here would make the hub retry a
      // handler that may have already half-performed the action.
      return {
        status: 200,
        body: {
          requestId: decision.requestId,
          status: "error",
          error: err instanceof Error ? err.message : String(err),
        },
      };
    }

    return { status: 200, body: result };
  };
}

File: sdk/go/stepauth/callback.go
package stepauth

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
)

// maxCallbackBodyBytes bounds a decision-delivery POST: only one small signed
// envelope is ever expected.
const maxCallbackBodyBytes = 1 << 20

// PendingState is what an SP must persist for a submitted request until its
// decision arrives (or it expires): the exact frozen request payload — to
// check the eventual decision's requestDigest — and which hub it was sent to,
// recorded at send time. JSON-serializable; the SDK persists nothing itself.
type PendingState struct {
	RequestID string `json:"requestId"`
	Hub       string `json:"hub"`
	HubEntity string `json:"hubEntity"`
	Payload   []byte `json:"payload"`

	// Result is the ExecutionResult already recorded for this request, if
	// the SP has executed it. Delivery is at-least-once, so a redelivery
	// must replay this rather than run the handler again.
	Result *ExecutionResult `json:"result,omitempty"`
}

// PendingLookup resolves a requestId to its stored PendingState: return the
// stored state for any requestId this SP issued, with Result set once the
// decision has been executed. Return (nil, nil) only for an id the SP never
// issued — never for an already-decided one, since nil reads to the hub as
// a rejection and it will retry delivery for 24h.
type PendingLookup func(ctx context.Context, requestID string) (*PendingState, error)

// DecisionHandler runs a verified decision (executing the approved action, or
// recording a denial) and reports back its execution. Two deliveries of one
// decision that arrive at once can both pass the recorded-result check before
// either records a result, so the handler must be idempotent by requestId.
type DecisionHandler func(ctx context.Context, decision *Decision) ExecutionResult

// NewCallbackHandler returns the SP's decision-delivery webhook: the hub
// POSTs a signed decision here directly, there is no polling. This endpoint
// carries NO transport auth of its own — the envelope signature IS the
// entire security, verified per VerifyDecision before handler ever runs.
func NewCallbackHandler(cfg *Config, lookup PendingLookup, handler DecisionHandler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			w.WriteHeader(http.StatusMethodNotAllowed)
			return
		}
		body, err := io.ReadAll(io.LimitReader(r.Body, maxCallbackBodyBytes))
		if err != nil {
			reject(w)
			return
		}

		// Decoded WITHOUT verifying, only to read requestId for routing; the
		// body stays untrusted until VerifyDecision runs below.
		payload, err := DecodeEnvelopePayload(body)
		if err != nil {
			reject(w)
			return
		}
		var probe struct {
			RequestID string `json:"requestId"`
		}
		if err := json.Unmarshal(payload, &probe); err != nil || probe.RequestID == "" {
			reject(w)
			return
		}

		// An unknown requestId is rejected before any verification is
		// attempted: never authenticate a decision the SP never asked for.
		pending, err := lookup(r.Context(), probe.RequestID)
		if err != nil || pending == nil {
			reject(w)
			return
		}

		decision, err := VerifyDecision(cfg, body, pending)
		if err != nil {
			reject(w)
			return
		}

		// A redelivery of an already-decided request must replay the
		// recorded result rather than run the handler again.
		if pending.Result != nil {
			respondJSON(w, *pending.Result)
			return
		}

		respondJSON(w, handler(r.Context(), decision))
	})
}

// respondJSON writes result as the 200 response body, or a rejection if it
// cannot be marshaled.
func respondJSON(w http.ResponseWriter, result ExecutionResult) {
	b, err := json.Marshal(result)
	if err != nil {
		reject(w)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(b)
}

// reject answers every rejection identically, regardless of cause.
func reject(w http.ResponseWriter) {
	w.WriteHeader(http.StatusUnauthorized)
}
